File: emailOrden.js
// emailOrden.js — ENVÍO DE ORDEN POR EMAIL (Zoho Mail API, NO BLOQUEANTE)
import { memoria } from "./index.js";
import PDFDocument from "pdfkit";
import fetch from "node-fetch";

/* ============================================================
   Helpers memoria
   ============================================================ */
function detectarModuloDesdeMemoria(idPago) {
  const spaces = ["trauma", "preop", "generales", "ia"];
  for (const s of spaces) {
    if (memoria.has(`${s}:${idPago}`)) return s;
  }
  return null;
}

function extraerEmail(datos) {
  if (!datos) return null;
  if (datos.email) return String(datos.email).trim();
  if (datos.traumaJSON?.paciente?.email)
    return String(datos.traumaJSON.paciente.email).trim();
  return null;
}

function emailValido(e) {
  return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(String(e || "").trim());
}

/* ============================================================
   PDF → Buffer
   ============================================================ */
async function generarPDFBuffer(datos, generador) {
  return new Promise((resolve, reject) => {
    const doc = new PDFDocument({ size: "A4", margin: 50 });
    const chunks = [];

    doc.on("data", (c) => chunks.push(c));
    doc.on("end", () => resolve(Buffer.concat(chunks)));
    doc.on("error", reject);

    generador(doc, datos);
    doc.end();
  });
}

/* ============================================================
   Zoho config
   ============================================================ */
const ZOHO_API = process.env.ZOHO_MAIL_API_DOMAIN;
const CLIENT_ID = process.env.ZOHO_CLIENT_ID;
const CLIENT_SECRET = process.env.ZOHO_CLIENT_SECRET;
const REFRESH_TOKEN = process.env.ZOHO_MAIL_REFRESH_TOKEN;

/* ============================================================
   Token refresh (interno, silencioso)
   ============================================================ */
async function refreshZohoToken() {
  try {
    const url =
      `https://accounts.zoho.com/oauth/v2/token` +
      `?grant_type=refresh_token` +
      `&client_id=${CLIENT_ID}` +
      `&client_secret=${CLIENT_SECRET}` +
      `&refresh_token=${REFRESH_TOKEN}`;

    const r = await fetch(url, { method: "POST" });
    const j = await r.json();

    if (!j.access_token) {
      console.error("❌ [ZOHO] No se pudo refrescar token:", j);
      return null;
    }

    process.env.ZOHO_MAIL_ACCESS_TOKEN = j.access_token;
    return j.access_token;
  } catch (e) {
    console.error("❌ [ZOHO] Error refresh token:", e?.message);
    return null;
  }
}

async function zohoFetch(url, options = {}) {
  let token = process.env.ZOHO_MAIL_ACCESS_TOKEN;

  let r = await fetch(url, {
    ...options,
    headers: {
      ...(options.headers || {}),
      Authorization: `Zoho-oauthtoken ${token}`,
    },
  });

  if (r.status !== 401) return r;

  token = await refreshZohoToken();
  if (!token) return r;

  return fetch(url, {
    ...options,
    headers: {
      ...(options.headers || {}),
      Authorization: `Zoho-oauthtoken ${token}`,
    },
  });
}

/* ============================================================
   ENVÍO DE CORREO — NUNCA BLOQUEANTE
   ============================================================ */
export async function enviarOrdenPorCorreo({ idPago, generadorPDF }) {
  try {
    console.log("📨 [ZOHO] Intento envío email. idPago:", idPago);

    const modulo = detectarModuloDesdeMemoria(idPago);
    if (!modulo) {
      console.warn("⚠️ [ZOHO] Módulo no detectado, se omite email");
      return;
    }

    const datos = memoria.get(`${modulo}:${idPago}`);
    if (!datos) {
      console.warn("⚠️ [ZOHO] Datos no encontrados, se omite email");
      return;
    }

    const email = extraerEmail(datos);
    if (!emailValido(email)) {
      console.warn("⚠️ [ZOHO] Email inválido, se omite:", email);
      return;
    }

    const pdfBuffer = await generarPDFBuffer(datos, generadorPDF);
    const pdfBase64 = pdfBuffer.toString("base64");

    // Obtener accountId
    const accResp = await zohoFetch(`${ZOHO_API}/mail/v1/accounts`);
    const accJson = await accResp.json();

    const accountId = accJson?.data?.[0]?.accountId;
    if (!accountId) {
      console.error("❌ [ZOHO] accountId no encontrado", accJson);
      return;
    }

    const asunto =
      modulo === "trauma"
        ? "Orden de imagenología – ICA"
        : modulo === "preop"
        ? "Orden preoperatoria – ICA"
        : modulo === "generales"
        ? "Orden de exámenes generales – ICA"
        : "Orden médica – ICA";

    const payload = {
      fromAddress: "contacto@icarticular.cl",
      toAddress: email,
      subject: asunto,
      content:
        "Estimado(a),\n\nAdjuntamos su orden médica generada por Asistencia ICA.\n\nInstituto de Cirugía Articular",
      attachments: [
        {
          fileName: "orden_medica.pdf",
          content: pdfBase64,
        },
      ],
    };

    const sendResp = await zohoFetch(
      `${ZOHO_API}/mail/v1/accounts/${accountId}/messages`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      }
    );

    const sendJson = await sendResp.json();

    if (!sendResp.ok) {
      console.error("❌ [ZOHO] Error envío:", sendJson);
      return;
    }

    console.log("📧 [ZOHO] Email enviado OK a:", email);
  } catch (e) {
    // 🔴 JAMÁS romper flujo PDF
    console.error("❌ [ZOHO] Error email (IGNORADO):", e?.message);
  }
}

File: correo/email_orden.py
# correo/email_orden.py
# Envío de orden por correo (Resend) — JAMÁS bloqueante
# Stateless — recibe datos completos, sin memoria server-side
# Depende de: httpx

import base64
import logging
import re
from typing import Callable

import httpx

logger = logging.getLogger("email_orden")

RESEND_API     = "https://api.resend.com/emails"
RESERVAS_BASE  = "https://reservas.icarticular.cl"

# Mapeo doctor_id (medicos.json) → professional_id (professionals.json ICA)
_MAP_DR_RESERVA = {
    "cristobal_huerta": "huerta",
    "jaime_espinoza":   "espinoza",
}


# ============================================================
# HELPERS
# ============================================================
def _email_valido(e: str) -> bool:
    return bool(re.match(r'^[^\s@]+@[^\s@]+\.[^\s@]+$', str(e or "").strip()))


def _extraer_email(datos: dict) -> str | None:
    if not datos:
        return None
    if datos.get("email"):
        return str(datos["email"]).strip()
    trauma_json = datos.get("traumaJSON") or {}
    paciente    = trauma_json.get("paciente") or {}
    if paciente.get("email"):
        return str(paciente["email"]).strip()
    return None


def _asunto_por_modulo(modulo: str) -> str:
    return {
        "trauma":    "Orden de imagenología – ICA",
        "preop":     "Orden preoperatoria – ICA",
        "generales": "Orden de exámenes – ICA",
    }.get(modulo, "Orden médica – ICA")


def _bloque_reserva(datos: dict) -> str:
    """
    Arma el bloque con link de reserva según derivación.
    No se envía si el paciente ya viene de una reserva.
    """
    # Si ya tiene reserva, no enviar link
    if datos.get("origen") == "reserva":
        return ""

    deriv  = datos.get("deriv") or {}
    doctor = deriv.get("doctor") or {}
    sede   = deriv.get("sede")   or {}

    doctor_id     = doctor.get("id") or ""
    doctor_nombre = doctor.get("nombre") or ""
    sede_nombre   = sede.get("nombre") or ""
    agenda        = doctor.get("agenda") or sede_nombre

    # Mapear al ID del sistema de reservas ICA
    reserva_id = _MAP_DR_RESERVA.get(doctor_id)
    if not reserva_id:
        return ""

    link = f"{RESERVAS_BASE}?dr={reserva_id}"

    lineas = [
        "─────────────────────────────────────",
        "📅 RESERVE SU HORA",
        "",
    ]

    if doctor_nombre:
        lineas.append(f"Médico recomendado: {doctor_nombre}")
    if agenda:
        lineas.append(f"Centro: {agenda}")

    lineas += [
        "",
        "Reserve su hora en línea:",
        link,
        "─────────────────────────────────────",
    ]

    return "\n".join(lineas)


# ============================================================
# ENVÍO — NUNCA BLOQUEANTE (captura todas las excepciones)
# ============================================================
async def enviar_orden_por_correo(
    datos:         dict,
    modulo:        str,
    generador_pdf: Callable[[dict], bytes],
    config:        dict,
) -> None:
    """
    datos:         dict con datos del paciente (debe incluir 'email')
    modulo:        'trauma' | 'preop' | 'generales' | 'ia'
    generador_pdf: callable(datos) → bytes
    config:        { resend_api_key, resend_from? }

    NUNCA lanza excepción — cualquier error se loggea y se ignora.
    """
    try:
        logger.info("📨 [RESEND] Intento envío email. módulo=%s", modulo)

        resend_key = config.get("resend_api_key") or ""
        if not resend_key:
            logger.warning("⚠️ [RESEND] RESEND_API_KEY no configurada")
            return

        from_addr = config.get("resend_from") or "contacto@icarticular.cl"

        email = _extraer_email(datos)
        if not _email_valido(email):
            logger.warning("⚠️ [RESEND] Email inválido: %s", email)
            return

        # — Generar PDF en memoria —
        try:
            pdf_bytes = generador_pdf(datos)
        except Exception as e:
            logger.error("❌ [RESEND] Error generando PDF: %s", e)
            return

        pdf_b64 = base64.b64encode(pdf_bytes).decode()
        asunto  = _asunto_por_modulo(modulo)

        # — Bloque de reserva (solo si no viene de reserva) —
        bloque_reserva = _bloque_reserva(datos)

        cuerpo = (
            "Estimado(a),\n\n"
            "Adjuntamos su orden médica generada por Asistencia ICA.\n"
        )

        if bloque_reserva:
            cuerpo += f"\n{bloque_reserva}\n"

        cuerpo += "\nInstituto de Cirugía Articular"

        payload = {
            "from":    f"Instituto de Cirugía Articular <{from_addr}>",
            "to":      [email],
            "subject": asunto,
            "text":    cuerpo,
            "attachments": [
                {
                    "filename": "orden_medica.pdf",
                    "content":  pdf_b64,
                }
            ],
        }

        async with httpx.AsyncClient(timeout=15.0) as client:
            r = await client.post(
                RESEND_API,
                headers={
                    "Authorization": f"Bearer {resend_key}",
                    "Content-Type":  "application/json",
                },
                json=payload,
            )
            r.raise_for_status()

        logger.info("📧 [RESEND] Email enviado OK a: %s", email)

    except Exception as e:
        # 🔴 NUNCA romper flujo PDF — solo loggear
        logger.error("❌ [RESEND] Error email (IGNORADO): %s", e)
        
